lib/types.js

```diff
@@ -196,11 +198,26 @@ class Type {
       let types = schema.map((obj) => {
         return Type.forSchema(obj, opts);
       });
+      let projectionFn;
       if (!UnionType) {
-        UnionType = isAmbiguous(types) ? WrappedUnionType : UnwrappedUnionType;
+        // either automatic detection or we have a projection function
+        if (typeof opts.wrapUnions === 'function') {
+          // we have a projection function
+          try {
+            projectionFn = opts.wrapUnions(types);
+            UnionType = typeof projectionFn !== 'undefined'
+              // projection function yields a function, we can use an Unwrapped type
+              ? UnwrappedUnionType
+              : WrappedUnionType;
+          } catch(e) {
+            throw new Error(`Error generating projection function: ${e}`);
+          }
```

Comment:
I noticed whilst testing that it is quite confusing for a user if the `wrapUnions` function producing a function and the produced function itself errors in the same executed codepath, hence I error early here on the first call, but wanted to make sure someone using that public interface understands better what the issue is, as otherwise it reads like an internal problem.
I am not hung up on it though, if you feel we shouldn't wrap and remove context, we can not wrap.

Reply:
OK. My current instinct is that users will be able to debug this from the stack trace directly (which is truncated if we wrap, making the underlying cause harder to find).

Let's avoid wrapping for now. We can revisit if/when we introduce an error representation which fully supports causes.

Reply:
9f6eaffec66396cd891e39eca93a41e71cf8d321